File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.Example1/CachingProxy.cs
namespace ProxyDesignPattern.Example1;

/// <summary>
/// CACHING PROXY - Caches video metadata to reduce server calls.
/// </summary>
public class CachingProxy : IVideoService
{
    private readonly IVideoService _innerService;
    private readonly Dictionary<string, Video> _cache = new();

    public CachingProxy(IVideoService innerService)
    {
        _innerService = innerService;
    }

    public Video? GetVideoInfo(string videoId)
    {
        // Check cache first
        if (_cache.TryGetValue(videoId, out var cachedVideo))
        {
            Console.WriteLine($"[CachingProxy] Cache HIT for video: {videoId}");
            return cachedVideo;
        }

        // Cache miss - fetch from service
        Console.WriteLine($"[CachingProxy] Cache MISS for video: {videoId}");
        var video = _innerService.GetVideoInfo(videoId);
        
        if (video != null)
        {
            _cache[videoId] = video;
            Console.WriteLine($"[CachingProxy] Video info cached: {videoId}");
        }

        return video;
    }

    public void StreamVideo(string videoId)
    {
        _innerService.StreamVideo(videoId);
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.Example1/src/CachingProxy.ts
import { IVideoService } from './IVideoService';
import { Video } from './Models';

// CACHING PROXY — caches video metadata
export class CachingProxy implements IVideoService {
  private readonly cache = new Map<string, Video>();

  constructor(private readonly innerService: IVideoService) {}

  getVideoInfo(videoId: string): Video | null {
    if (this.cache.has(videoId)) {
      console.log(`[CachingProxy] Cache HIT for video: ${videoId}`);
      return this.cache.get(videoId)!;
    }
    console.log(`[CachingProxy] Cache MISS for video: ${videoId}`);
    const video = this.innerService.getVideoInfo(videoId);
    if (video) {
      this.cache.set(videoId, video);
      console.log(`[CachingProxy] Video info cached: ${videoId}`);
    }
    return video;
  }

  streamVideo(videoId: string): void {
    this.innerService.streamVideo(videoId);
  }
}
